overlord/healthstate/healthstate.go

```diff
@@ -0,0 +1,192 @@
+	OkayStatus
+	WaitingStatus
+	BlockedStatus
+	ErrorStatus
+)
+
+var knownStatuses = []string{"unknown", "okay", "waiting", "blocked", "error"}
+
+func StatusLookup(str string) (HealthStatus, error) {
+	for i, k := range knownStatuses {
+		if k == str {
+			return HealthStatus(i), nil
+		}
+	}
+	return -1, fmt.Errorf("invalid status %q, must be one of %s", str, strutil.Quoted(knownStatuses))
+}
+
+func (s HealthStatus) String() string {
+	if s < 0 || s >= HealthStatus(len(knownStatuses)) {
+		return fmt.Sprintf("invalid (%d)", s)
+	}
+	return knownStatuses[s]
+}
+
+type HealthState struct {
+	Revision  snap.Revision `json:"revision"`
+	Timestamp time.Time     `json:"timestamp"`
+	Status    HealthStatus  `json:"status"`
+	Message   string        `json:"message,omitempty"`
+	Code      string        `json:"code,omitempty"`
+}
+
+func Init(hookManager *hookstate.HookManager) {
+	hookManager.Register(regexp.MustCompile("^check-health$"), newHealthHandler)
+}
+
+func newHealthHandler(ctx *hookstate.Context) hookstate.Handler {
+	return &healthHandler{context: ctx}
+}
+
+type healthHandler struct {
+	context *hookstate.Context
+}
+
+// Before is called just before the hook runs -- nothing to do beyond setting a marker
+func (h *healthHandler) Before() error {
+	// we use the 'health' entry as a marker to not add OnDone to
+	// the snapctl set-health execution
+	h.context.Lock()
+	h.context.Set("health", struct{}{})
+	h.context.Unlock()
+	return nil
+}
+
+func (h *healthHandler) Done() error {
+	var health HealthState
+
+	h.context.Lock()
+	err := h.context.Get("health", &health)
+	h.context.Unlock()
+
+	if err != nil && err != state.ErrNoState {
+		// note it can't actually be state.ErrNoState because Before sets it
+		// (but if it were, health.Timestamp would still be zero)
+		return err
+	}
+	if health.Timestamp.IsZero() {
+		// health was actually the marker (or err == state.ErrNoState)
+		health = HealthState{
+			Revision:  h.context.SnapRevision(),
+			Timestamp: time.Now(),
+			Status:    UnknownStatus,
+			Code:      "snapd-hook-no-health-set",
+			Message:   "hook did not call set-health",
+		}
+	}
+
+	return h.appendHealth(&health)
+}
+
+func (h *healthHandler) Error(err error) error {
+	return h.appendHealth(&HealthState{
+		Revision:  h.context.SnapRevision(),
+		Timestamp: time.Now(),
+		Status:    UnknownStatus,
+		Code:      "snapd-hook-failed",
+		Message:   "hook failed",
+	})
+}
+
+func (h *healthHandler) appendHealth(health *HealthState) error {
+	st := h.context.State()
+	st.Lock()
+	defer st.Unlock()
+
+	return appendHealth(h.context, health)
+}
+
+func appendHealth(ctx *hookstate.Context, health *HealthState) error {
+	st := ctx.State()
+
+	var hs map[string]*HealthState
+	if err := st.Get("health", &hs); err != nil {
+		if err != state.ErrNoState {
+			return err
+		}
+		hs = map[string]*HealthState{}
+	}
+	hs[ctx.InstanceName()] = health
+	st.Set("health", hs)
+
+	return nil
+}
+
+func SetHealth(ctx *hookstate.Context) error {
```

Comment:
this doesn't seem to be unit tested

Reply:
I would also call it SetHealthFromHookContext,  I would expect a different signature for something just called SetHealth, (it would take the HealthState directly I think)

Reply:
okay